Contributing: PRs: Document how and why to run static checks

grahamwhaley opened the pull request:
We need to encourage developers to run the static checks locally
before submitting PRs, or they PRs will just bounce on the first
go. We also need to document how to run the checks, and any other
caveats.

Fixes: #45

Signed-off-by: Graham Whaley <graham.whaley@intel.com>

grahamwhaley commented:
Marking as WIP, as it seems a touch more complex than this maybe:
 - we also have 'make test', which looks like it requires cni binaries to be installed
 - 'make check' seems to hang up at some point for me, so I need to figure that out

grahamwhaley commented:
The hang is fixed with PR https://github.com/kata-containers/runtime/pull/479.

klynnrif commented on CONTRIBUTING.md:
```diff
@@ -64,6 +65,32 @@ issue yourself. This gives the team visibility of the work you plan to do.
+
+Kata Containers utilizes a number of [CI systems](https://en.wikipedia.org/wiki/Continuous_integration)
+to automatically check every PR. As well as performing execution tests locally on
+your PR before submission, you are also strongly encouraged to run the suite of
```
Removing "also": your PR before submission, you are strongly encouraged to run the suite of

klynnrif commented on CONTRIBUTING.md:
```diff
@@ -64,6 +65,32 @@ issue yourself. This gives the team visibility of the work you plan to do.
+your PR before submission, you are also strongly encouraged to run the suite of
+[static checks](https://github.com/kata-containers/tests/blob/master/.ci/static-checks.sh)
+the CI will execute. The easiest way to execute these in most of the repositories is
+to invoke the `Makefile` `check` rule, whilst enabling developer mode:
```
Suggested rewrite (unless I've changed the meaning): 
to invoke the `Makefile` `check` rule, while developer mode is enabled:

klynnrif commented on CONTRIBUTING.md:
````diff
@@ -64,6 +65,32 @@ issue yourself. This gives the team visibility of the work you plan to do.
+$ make check
+```
+
+Execution of this should complete with no errors. If errors are reported, fix them
````
Lines 82-83 suggested rewrite: 
This should execute without errors. If errors are reported, fix them before you submit your PR.

klynnrif commented on CONTRIBUTING.md:
```diff
@@ -64,6 +65,32 @@ issue yourself. This gives the team visibility of the work you plan to do.
+> **Note:** To ensure you replicate the static checks performed by the CI system, it is
+recommended that you:
+> * Ensure you have a 'clean' source tree, as the checks will check all files present, and
+may fail if you have out of date or extra files in your tree.
```
might fail if you have extra files or your files are out of date in your tree.

klynnrif commented on CONTRIBUTING.md:
```diff
@@ -64,6 +65,32 @@ issue yourself. This gives the team visibility of the work you plan to do.
+recommended that you:
+> * Ensure you have a 'clean' source tree, as the checks will check all files present, and
+may fail if you have out of date or extra files in your tree.
+> * Ensure you have an up-to date (or have not installed, as the static check scripts will
```
Lines 89-92 suggested rewrite (hopefully I did not change meaning): 
* Ensure that [gometalinter](https://github.com/alecthomas/gometalinter) is current or has not been installed due to the fact that the static check scripts will install. Changes in either the linters used or the Kata Containers code base can produce spurious errors that do not fail inside the CI systems.

jodh-intel commented on CONTRIBUTING.md:
```diff
@@ -64,6 +65,32 @@ issue yourself. This gives the team visibility of the work you plan to do.
+Kata Containers utilizes a number of [CI systems](https://en.wikipedia.org/wiki/Continuous_integration)
+to automatically check every PR. As well as performing execution tests locally on
+your PR before submission, you are also strongly encouraged to run the suite of
+[static checks](https://github.com/kata-containers/tests/blob/master/.ci/static-checks.sh)
```
I totally agree that we need this sort of helpful doc but I'm wondering how much should be here and how much should be in https://github.com/kata-containers/tests/blob/master/.ci/README.md (with a link to it from here)?

It might be worth mentioning that you can now run `.ci/static-checks.sh -h`as there are a few options (which can help speed up debugging as you can run just a single category of checks if you wish).

grahamwhaley commented:
Right - yes, this is still WIP whilst I ironed out the wrinkles like the still pending https://github.com/kata-containers/runtime/pull/496
I'll try to re-visit - may have to be back end of the week though.

bergwolf commented:
@grahamwhaley Thanks! I myself always wonder how I could do this check locally :)

One question though, how much memory do I need to run it? I just tried `make check` with `KATA_DEV_MODE=true` on a 4GB memory VM but it got OOM-killed.
```
WARNING: failed to execute linter errcheck -abspath .: fork/exec /golang/bin/errcheck: cannot allocate memory
Makefile:397: recipe for target 'check-go-static' failed
make: *** [check-go-static] Error 3
```
And I'm on a clean runtime tree:
```
[macbeth@runtime]$git status
On branch qemu
nothing to commit, working tree clean
```

Do you know which command is consuming so much memory? Can I disable it when running locally or can we somehow reduce its memory footprint?

jodh-intel commented:
It looks like one of the `gometalinter` checks is the culprit. I don't think those checkers allow you to specify a maximum amount of memory to use but you could add an option to the static checks script to enable/disable the current linters we enable maybe?:

- https://github.com/kata-containers/tests/blob/master/.ci/static-checks.sh#L250..L264

jodh-intel commented:
@grahamwhaley - https://github.com/kata-containers/runtime/pull/496 is now merged so can we drop the WIP label?

Looks like you still need to update the doc element of this PR though?

jodh-intel commented:
Yep - doc changes still need updating fwics.

grahamwhaley commented:
Updated. @jodh-intel @klynnrif please re-check.
 - rebased
 - updated with the feedback - thx
 - added 'make tests' to the mix, as we should do that as well as 'check'
 - dropped the WIP

grahamwhaley commented:
still looking for review from one more @kata-containers/community member ;-)

jodh-intel commented:
Ping @bergwolf, @gnawux, @WeiZhang555 - could you take a quick look at this small doc change?

caoruidong commented:
Ping @bergwolf @gnawux @WeiZhang555
